File: src/lib/utils/invites.ts
import { randomBytes } from 'node:crypto';

const INVITE_ALPHABET = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
// 32-char alphabet over 6 positions = 32^6 = ~10^9 combinations. With per-IP
// signup rate limiting and a 7-day TTL on each code, brute-forcing an active
// code is impractical even on a shared instance. Bumped from 4 chars (32^4 ≈
// 1M combos) which felt thin for a public deployment.
const INVITE_CODE_LENGTH = 6;

export function generateInviteCodeRaw(): string {
  const bytes = randomBytes(INVITE_CODE_LENGTH);
  let suffix = '';
  for (let i = 0; i < INVITE_CODE_LENGTH; i++) {
    suffix += INVITE_ALPHABET[bytes[i] % INVITE_ALPHABET.length];
  }
  return `BEBE-${suffix}`;
}

export function isValidInviteCodeFormat(code: string): boolean {
  const pattern = new RegExp(`^BEBE-[A-Z2-9]{${INVITE_CODE_LENGTH}}$`);
  return pattern.test(code) && !/[0OI1]/.test(code);
}

File: phoenix/lib/diversif/utils/invites.ex
defmodule Diversif.Utils.Invites do
  @moduledoc """
  Invite code generator. Mirrors `src/lib/utils/invites.ts`.

  Alphabet excludes ambiguous chars (0, O, I, 1). 6 chars over 32 = ~10^9 combos;
  with a 7-day TTL and signup rate limiting, brute force is impractical.
  """

  @alphabet ~c"ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
  @code_length 6
  @legacy_length 4

  @spec generate_code() :: String.t()
  def generate_code do
    suffix =
      for <<b <- :crypto.strong_rand_bytes(@code_length)>>,
          into: "",
          do: <<Enum.at(@alphabet, rem(b, length(@alphabet)))>>

    "BEBE-" <> suffix
  end

  @spec valid_format?(String.t()) :: boolean()
  def valid_format?(code) when is_binary(code) do
    pattern =
      Regex.compile!("^BEBE-([A-Z2-9]{#{@legacy_length}}|[A-Z2-9]{#{@code_length}})$")

    Regex.match?(pattern, code) and not String.match?(code, ~r/[0OI1]/)
  end

  def valid_format?(_), do: false
end
